mne/decoding/ssd.py

```diff
@@ -0,0 +1,316 @@
+        As in :func:`mne.decoding.SPoC`
+        The default is None.
+    rank : None | dict | ‘info’ | ‘full’
+        As in :func:`mne.decoding.SPoC`
+        This controls the rank computation that can be read from the
+        measurement info or estimated from the data.
+        See Notes of :func:`mne.compute_rank` for details.
+        We recommend to use 'full' when working with epoched data.
+
+    Attributes
+    ----------
+    filters_ : array, shape(n_channels, n_components)
+        The spatial filters to be multiplied with the signal.
+    patterns_ : array, shape(n_components, n_channels)
+        The patterns for reconstructing the signal from the filtered data.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+
+    def __init__(self, info, filt_params_signal, filt_params_noise,
+                 reg=None, n_components=None, picks=None,
+                 sort_by_spectral_ratio=True, return_filtered=False,
+                 n_fft=None, cov_method_params=None, rank=None):
+        """Initialize instance."""
+        dicts = {"signal": filt_params_signal, "noise": filt_params_noise}
+        for param, dd in [('l', 0), ('h', 0), ('l', 1), ('h', 1)]:
+            key = ('signal', 'noise')[dd]
+            if param + '_freq' not in dicts[key]:
+                raise ValueError(
+                    '%s must be defined in filter parameters for %s'
+                    % (param + '_freq', key))
+            val = dicts[key][param + '_freq']
+            if not isinstance(val, (int, float)):
+                raise ValueError(
+                    'Frequencies must be numbers, got %s' % type(val))
+        # check freq bands
+        if (filt_params_noise['l_freq'] > filt_params_signal['l_freq'] or
+                filt_params_signal['h_freq'] > filt_params_noise['h_freq']):
+            raise ValueError('Wrongly specified frequency bands!\n'
+                             'The signal band-pass must be within the noise '
+                             'band-pass!')
+        ch_types = {channel_type(info, ii)
+                    for ii in range(info['nchan'])}
+        if len(ch_types) > 1:
+            raise ValueError('At this point SSD only supports fitting '
+                             'single channel types. Your info has %i types' %
+                             (len(ch_types)))
+        self.info = info
+        self.freqs_signal = (filt_params_signal['l_freq'],
+                             filt_params_signal['h_freq'])
+        self.freqs_noise = (filt_params_noise['l_freq'],
+                            filt_params_noise['h_freq'])
+        self.filt_params_signal = filt_params_signal
+        self.filt_params_noise = filt_params_noise
+        self.sort_by_spectral_ratio = sort_by_spectral_ratio
+        if n_fft is None:
+            self.n_fft = int(self.info['sfreq'])
+        else:
+            self.n_fft = int(n_fft)
+        self.picks_ = (Ellipsis if picks is None else picks)
```

Comment:
Move this to the top and do
```suggestion
        self.picks_ = _picks_to_idx(info, picks, none='data', exclude='bads')
        del picks
```
then use `self.picks_` in the `_get_channel_types` above